src/realm/backup_restore.cpp

```diff
@@ -0,0 +1,202 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include <realm/backup_restore.hpp>
+#include <realm/util/file.hpp>
+
+#include <vector>
+#include <chrono>
+
+namespace realm {
+
+/*
+ * IMPORTANT: The following two arrays must be kept updated
+ * as new versions are released or if rollback is ever done.
+ */
+
+using version_list_t = BackupHandler::version_list_t;
+using version_time_list_t = BackupHandler::version_time_list_t;
+
+// Note: accepted versions should have new versions added at front
+version_list_t BackupHandler::accepted_versions_ = {20, 11, 10, 9, 8, 7, 6, 0};
+
+// the pair is <version, age-in-seconds>
+// we keep backup files in 3 months.
+static constexpr int three_months = 3 * 31 * 24 * 60 * 60;
+version_time_list_t BackupHandler::delete_versions_{{20, three_months}, {11, three_months}, {10, three_months},
+                                                    {9, three_months},  {8, three_months},  {7, three_months},
+                                                    {6, three_months}};
+
+
+// helper functions
+std::string backup_name(std::string prefix, int version)
+{
+    return prefix + "v" + std::to_string(version) + ".backup.realm";
+}
+
+bool backup_exists(std::string prefix, int version)
+{
+    std::string fname = backup_name(prefix, version);
+    return util::File::exists(fname);
+}
+
+std::string BackupHandler::get_prefix_from_path(const std::string& path)
+{
+    auto size = path.size();
+
+    // remove a suffix ".realm" but add back the "."
+    if (size > 6 && path.substr(size - 6, 6) == ".realm") {
+        return path.substr(0, size - 5); // include '.'
+    }
+
+    // if no ".realm" suffix, at least ensure a terminating "."
+    if (path[size - 1] == '.') {
+        return path;
+    }
+    return path + ".";
+}
+
+BackupHandler::BackupHandler(const std::string& path, const version_list_t& accepted,
+                             const version_time_list_t& to_be_deleted)
+{
+    m_path = path;
+    m_prefix = get_prefix_from_path(path);
+    m_accepted_versions = accepted;
+    m_delete_versions = to_be_deleted;
+}
+
+bool BackupHandler::must_restore_from_backup(int current_file_format_version)
+{
+    if (current_file_format_version == 0)
+        return false;
+    if (is_accepted_file_format(current_file_format_version))
+        return false;
+    for (auto v : m_accepted_versions) {
+        if (backup_exists(m_prefix, v))
+            return true;
+    }
+    return false;
+}
+
+bool BackupHandler::is_accepted_file_format(int version)
+{
+    for (auto v : m_accepted_versions) {
+        if (v == version)
+            return true;
+    }
+    return false;
+}
+
+void BackupHandler::restore_from_backup()
+{
+    for (auto v : m_accepted_versions) {
+        if (backup_exists(m_prefix, v)) {
+            prep_logging();
```

Comment:
Should we be calling this inside the loop?

Reply:
Yeah, that's optimal.